pkg/target/hostMTCP.go

```diff
@@ -0,0 +1,362 @@
+			break
+		}
+
+		h.Lg.Warn("error sending value to host. Reconnect and retry..", zap.Error(err))
+		err = conn.Close()
+		if err != nil {
+			// not retrying here, file descriptor may be lost
+			h.Lg.Error("error closing the connection", zap.Error(err))
+		}
+	}
+}
+
+// Flush periodically flushes the buffer and performs a write.
+func (h *HostMTCP) flush(d time.Duration) {
+	t := time.NewTicker(d)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-h.stop:
+			return
+		case <-t.C:
+			h.LockMTCP()
+			h.tryToFlushIfNecessary()
+			h.UnlockMTCP()
+		}
+	}
+}
+
+// Requires h.Conn.Mutex lock.
+func (h *HostMTCP) tryToFlushIfNecessary() {
+	for c := range h.MTCPs {
+		if h.MTCPs[c].W != nil && h.MTCPs[c].W.Buffered() != 0 {
+			if h.MTCPs[c].Conn == nil {
+				h.ensureConnection(&h.MTCPs[c])
+			} else {
+				h.keepConnectionFresh(&h.MTCPs[c])
+			}
+			err := h.MTCPs[c].W.Flush()
+			if err != nil {
+				h.Lg.Error("error while flushing the host buffer", zap.Error(err), zap.String("target_name", h.Name), zap.Uint16("target_port", h.Port))
+				h.MTCPs[c].Conn = nil
+				h.MTCPs[c].W = nil
+			}
+			h.MTCPs[c].LastConnUse = time.Now()
+		}
+	}
+}
+
+// Requires h.Conn.Mutex lock.
+// This function may take a long time.
+func (h *HostMTCP) keepConnectionFresh(conn *MultiConnection) {
+	// 0 value = don't refresh connections
+	if h.conf.TCPOutConnectionRefreshPeriodSec != 0 {
+		if conn != nil && (time.Since(conn.LastConnUse) > time.Second*time.Duration(h.conf.TCPOutConnectionRefreshPeriodSec)) {
+			h.oldConnectionRefresh.Inc()
+			h.oldConnectionRefreshTotal.Inc()
+
+			err := conn.Close()
+			if err != nil {
+				h.Lg.Error("closing connection to target host failed", zap.String("host", h.Name))
+			}
+			h.ensureConnection(conn)
+		}
+	}
+}
+
+// Tries to connect as long as Host.Conn.Conn == nil.
+// Requires h.Conn.Mutex lock.
+// This function may take a long time.
+func (h *HostMTCP) ensureConnection(conn *MultiConnection) {
+	for waitMs, attemptCount := uint32(0), 1; conn.Conn == nil; {
+		if h.jitterEnabled {
+			jitterAmplitude := waitMs / 2
+			if h.minJitterAmplitudeMs > jitterAmplitude {
+				jitterAmplitude = h.minJitterAmplitudeMs
+			}
+			waitMs = waitMs/2 + uint32(h.rnd.Intn(int(jitterAmplitude*2)))
+		}
+
+		time.Sleep(time.Duration(waitMs) * time.Millisecond)
+
+		if waitMs < h.conf.MaxHostReconnectPeriodMs {
+			waitMs = waitMs*2 + h.conf.HostReconnectPeriodDeltaMs
+		}
+		if waitMs >= h.conf.MaxHostReconnectPeriodMs {
+			waitMs = h.conf.MaxHostReconnectPeriodMs
+		}
+
+		h.connect(conn, attemptCount)
+		attemptCount++
+	}
+}
+
+// Connect multiple TCP connects to target host via TCP. If unsuccessful, sets conn to nil.
+// Requires h.Conn.Mutex lock.
+func (h *HostMTCP) connect(c *MultiConnection, attemptCount int) {
+	conn, err := h.getConnectionToHost()
+	if err != nil {
+		h.Lg.Warn("connection to target host failed")
+		c.Conn = nil
+		if attemptCount == 1 {
+			if h.Available.Load() {
+				h.setAvailability(false)
```

Comment:
Probably it is not a deal breaker but semanticly the host should be set as unavailable if all connections fail to establish. Now every failure in connect() on each of the MultiiConnection connections causes the host to be unavailable, if I'm not mistaken.

Reply:
Agreed, but it's better to be safe than sorry here. Let me do that explicitly.